api/admin/controller/individual_admin_settings.py

```diff
@@ -57,25 +54,58 @@ def process_post(self):
         else:
             roles = []
 
-        roles_error = self.handle_roles(user, admin, roles, settingUp)
+        roles_error = self.handle_roles(admin, roles, settingUp)
         if roles_error:
             return roles_error
 
         password = flask.request.form.get("password")
-        self.handle_password(password, admin, is_new, user)
+        self.handle_password(password, admin, is_new, settingUp)
 
         return self.response(admin, is_new)
 
     def check_permissions(self, admin, settingUp):
         """Before going any further, check that the user actually has permission
          to create/edit this type of admin"""
-        # User must be a sitewide library manager in order to create/edit
-        # a sitewide library manager.
-        if admin.is_sitewide_library_manager() and not settingUp:
-            self.require_sitewide_library_manager()
-        # User must be a system admin in order to create/edit a system admin.
-        if admin.is_system_admin() and not settingUp:
-            self.require_system_admin()
+
+         # For readability: the person who is submitting the form is referred to as "user"
+         # rather than as something that could be confused with "admin" (the admin
+         # which the user is submitting the form in order to create/edit.)
+
+        if not settingUp:
+            user = flask.request.admin
+            library = None
+            if admin.roles:
+                library = admin.roles[0].library
+
+            # System admin has all permissions.
+            if user.is_system_admin():
+                return
+
+            # If we've hit this point, then the user isn't a system admin.  If the
+            # admin is a system admin, the user won't be able to do anything.
+            if admin.is_system_admin():
+                raise AdminNotAuthorized
+
+            # By this point, we know no one is a system admin.
+            if user.is_sitewide_library_manager():
+                return
+
+            # The user isn't a system admin or a sitewide manager.
+            if admin.is_sitewide_library_manager():
+                raise AdminNotAuthorized
+
+            # By process of elimination, the admin must be either a librarian or
+            # a (non-sitewide) library manager.
+
+            # Is the user a library manager?
+            user_manages_admin_library = library and user.is_library_manager(library)
+            manager_role = (not library and filter(lambda role: role.role == AdminRole.LIBRARY_MANAGER, user.roles))
+            user_is_library_manager = (user_manages_admin_library or manager_role)
```

Comment:
I think there are some cases where this code won't work. For example: admin is a librarian of library A and library B, and user is a library manager of library B. User tries to remove admin's librarianship from library B only, but gets AdminNotAuthorized because they aren't a manager of library A.

Reply:
I think there is other code that already checks role-related permissions so I would just remove this from here.